### PR TITLE
Add new SafeHandleMarshaller type to provide out-of-generator marshalling support.

### LINKED ISSUE #74035: [API Proposal]: Marshaller type to support SafeHandle in source-generated marshalling without hard-coding in the generator
### Background and motivation

Today, we support marshalling `System.Runtime.InteropServices.SafeHandle`-derived types intrinsically in the interop source generators. This is one of the only types that we do this for since we moved string marshalling out of the generator during the .NET 7 development cycle. Additionally, `SafeHandle` marshalling is extremely complex, much more than any other built-in type.

As a result, we'd like to introduce a marshaller type so we can move the interop source generators to use this type instead of emitting the code directly.

This change will make the interop code more serviceable, since we can't service source-generated code.

### API Proposal

```csharp
namespace System.Runtime.InteropServices.Marshalling;

[CustomMarshaller(typeof(CustomMarshallerAttribute.GenericPlaceholder), MarshalMode.ManagedToUnmanagedIn, typeof(ManagedToUnmanagedIn))]
[CustomMarshaller(typeof(CustomMarshallerAttribute.GenericPlaceholder), MarshalMode.ManagedToUnmanagedRef, typeof(ManagedToUnmanagedRef))]
[CustomMarshaller(typeof(CustomMarshallerAttribute.GenericPlaceholder), MarshalMode.ManagedToUnmanagedOut, typeof(ManagedToUnmanagedOut))]
public static class SafeHandleMarshaller<T>
    where T : System.Runtime.InteropServices.SafeHandle
{
    public struct ManagedToUnmanagedIn
    {
        public void FromManaged(T handle);

        public IntPtr ToUnmanaged();

        public void Free();
    }

    public struct ManagedToUnmanagedRef
    {
        public ManagedToUnmanagedRef();

        public void FromManaged(T handle);

        public IntPtr ToUnmanaged();

        public void FromUnmanaged(IntPtr value);

        public T ToManagedGuaranteed();

        public void OnInvoked();

        public void Free();
    }

    public struct ManagedToUnmanagedOut
    {
        public ManagedToUnmanagedOut();

        public void FromUnmanaged(IntPtr value);

        public T ToManaged();

        public void OnInvoked();

        public void Free();
    }
}

```

### API Usage

```csharp
// Marshaller will be implicitly used
[LibraryImport("foo")]
public void Method(Microsoft.Win32.SafeFileHandle handle);
[LibraryImport("foo")]
public void Method2([MarshalUsing(typeof(SafeHandleMarshaller<Microsoft.Win32.SafeFileHandle>))] Microsoft.Win32.SafeFileHandle handle);
```

### Alternative Designs

We considered adding a `new()` constraint to the generic parameter to enforce our requirements around constructing new instances of the SafeHandle-derived type for the ref and out marshallers; however, this constraint would cause breaking changes for APIs that take non-concrete SafeHandle types as by-value or by-ref `in` parameters.

### Risks

_No response_

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Fixes #74035

We can't remove the built-in marshalling support from the generator yet, but once the out-of-band packages we ship don't support .NET 6. we can remove the built-in support that emits the marshalling code in the stub. I believe the .NET 9 packages won't support .NET 6, so once we snap for .NET 9 and update how we ship the packages, we can clean this up.

This PR also adds a requested feature to the SafeHandle marshaller: If the call to native code fails, we'll call `Dispose()` on the pre-allocated handle to avoid leaking it to the finalizer queue.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/SafeHandleMarshaller.cs [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+            /// Free any resources and reference counts owned by the marshaller.
+            /// </summary>
+            public void Free()
+            {
+                // If we never invoked the call, then we aren't going to use the
+                // new handle. Dispose it now to avoid clogging up the finalizer queue
+                // unnecessarily.
+                if (!_callInvoked)
+                {
+                    _newHandle!.Dispose();
```
Comment (teo-tsirpanis) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/SafeHandleMarshaller.cs:194]: Would `SetHandleAsInvalid` also work?

Comment (jkoritzinsky) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/SafeHandleMarshaller.cs:194]: In this case, the handle is already invalid, as it was default-constructed. We specifically want to dispose it here so we get the behavior of `Dispose` (specifically that the handle object can't be un-disposed). There's not much difference between the two operations, but I'd prefer to use `Dispose`.

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/SafeHandleMarshallingInfoProvider.cs [hunk 2]
```diff
@@ -47,18 +51,50 @@ public bool CanProvideMarshallingInfoForType(ITypeSymbol type)
...
+                        hasDefaultConstructor = true;
                         hasAccessibleDefaultConstructor = _compilation.IsSymbolAccessibleWithin(ctor, _containingScope);
                         break;
                     }
                 }
             }
+
+            if (_safeHandleMarshallerType is not null)
+            {
+
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/SafeHandleMarshallingInfoProvider.cs:71]:
```suggestion
```

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/SafeHandleMarshallingInfoProvider.cs [hunk 3]
```diff
@@ -47,18 +51,50 @@ public bool CanProvideMarshallingInfoForType(ITypeSymbol type)
...
                     if (ctor.Parameters.Length == 0)
                     {
+                        hasDefaultConstructor = true;
                         hasAccessibleDefaultConstructor = _compilation.IsSymbolAccessibleWithin(ctor, _containingScope);
                         break;
                     }
                 }
             }
+
+            if (_safeHandleMarshallerType is not null)
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/SafeHandleMarshallingInfoProvider.cs:69]: Let's flip this predicate and using the current return statement here.
```suggestion
            if (_safeHandleMarshallerType is null)
            {
                return new SafeHandleMarshallingInfo(hasAccessibleDefaultConstructor, type.IsAbstract);
            }
```

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/SafeHandleMarshaller.cs [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+            private bool _callInvoked;
+            private T _newHandle;
+
+            /// <summary>
+            /// Create the marshaller in a default state.
+            /// </summary>
+            public ManagedToUnmanagedOut()
+            {
+                _callInvoked = false;
+                _newHandle = (T)Activator.CreateInstance(typeof(T), nonPublic: true)!;
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/SafeHandleMarshaller.cs:158]: Please add a comment here as to the unfortunate history around non-public constructors and `SafeHandle`.

**Review by AaronRobinsonMSFT — COMMENTED** [review]
Do we need any additional testing here or is this simply making our existing semantics public? Please do a quick audit just to make sure we have good coverage here.

**Comment by AaronRobinsonMSFT** [comment]
It would be good to see if we have a test for this marshaller and arrays. Hopefully we are close to it just being apart of the system and that should be naturally covered, but that is a scenario that is troublesome and very bad when we get it wrong.

**Comment by jkoritzinsky** [comment]
Explicitly did not support marshalling arrays of SafeHandles beforehand and now we won't by construction of our custom type marshalling system. I'll add some tests in CompileFails to validate.

**Review by AaronRobinsonMSFT — DISMISSED** [review]

**Comment by jkoritzinsky** [comment]
Looks like adding this marshaller might have caused a significant size regression in NativeAOT, though I'm not sure why the regression would be as large as it is. I'll try to investigate that today.

**Comment by jkoritzinsky** [comment]
Found the problem, and it's not good. NativeAOT doesn't accelerate the `Activator.CreateInstance(Type, bool)` method, so having the marshaller use that ends up being extremely expensive as it pulls in the whole reflection stack. I'll try to figure out something here.

**Comment by jkoritzinsky** [comment]
Marking as blocked as we need to resolve how to avoid rooting the reflection stack before we merge this in.

**Comment by jkoritzinsky** [comment]
Based on a conversation with @jkotas in #85541, we have decided to take a breaking change to only allow `SafeHandle`-derived types with `public` constructors in `ref`, `out`, and `return` scenarios. That way our source-generated marshalling rules are following our guidance and the `SafeHandleMarshaller` type does not pull in the reflection stack.

**Comment by f2bo** [comment]
How would an array be handled using this marshaller? For example,

```C#
[LibraryImport("some_lib")]
public static partial void WaitForMultipleSignals([MarshalUsing(typeof(ArrayMarshaller<,>), CountElementName = nameof(numSignals))] SafeWaitHandle[] signals, int numSignals);
```

Wouldn't it also require support for `ElementIn` and `ElementOut` marshal modes, or is there some other mechanism to handle this scenario?

**Comment by AaronRobinsonMSFT** [comment]
> How would an array be handled using this marshaller?

We don't support arrays of `SafeHandle`s.

> Wouldn't it also require support for ElementIn and ElementOut marshal modes

Yes. We are not supporting this at present.

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/SafeHandleMarshaller.cs [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+            /// </summary>
+            public void Free()
+            {
+                if (_addRefd)
+                {
+                    _handle!.DangerousRelease();
+                }
+            }
+        }
+
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/SafeHandleMarshaller.cs:52]:
```suggestion
```

Comment (AaronRobinsonMSFT) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/SafeHandleMarshaller.cs:52]: Extra line.

File: src/libraries/System.Runtime.InteropServices/tests/LibraryImportGenerator.UnitTests/CompileFails.cs [hunk 6]
```diff
@@ -113,6 +113,9 @@ public static IEnumerable<object[]> CodeSnippetsToCompile()
             // Abstract SafeHandle type by reference
             yield return new object[] { ID(), CodeSnippets.BasicParameterWithByRefModifier("ref", "System.Runtime.InteropServices.SafeHandle"), 1, 0 };
 
+            // SafeHandle array
+            yield return new object[] { ID(), CodeSnippets.MarshalAsArrayParametersAndModifiers("Microsoft.Win32.SafeHandles.SafeFileHandle"), 5, 0 };
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/tests/LibraryImportGenerator.UnitTests/CompileFails.cs:117]: Do we want a compile fails for a `private` constructor?

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/SafeHandleMarshallingInfoProvider.cs [hunk 7]
```diff
@@ -47,19 +51,53 @@ public bool CanProvideMarshallingInfoForType(ITypeSymbol type)
...
+                entryPointType,
+                type,
+                _compilation,
+                out CustomTypeMarshallers? marshallers))
+            {
+                return NoMarshallingInfo.Instance;
+            }
+
+            // If the SafeHandle-derived type doesn't have a default constructor or is abstract,
+            // we only support managed-to-unmanaged marshalling
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/SafeHandleMarshallingInfoProvider.cs:88]: Do we need to update an analyzer or fixer in this case?

**Review by AaronRobinsonMSFT — COMMENTED** [review]
I think we should also update our compat doc about `SafeHandle` scenarios. This is breaking change from `DllImport`, which we've typically documented in the design doc.

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/SafeHandleMarshallingInfoProvider.cs (hunk 7 above)
Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/SafeHandleMarshallingInfoProvider.cs:88]: We already have an analyzer and fixer in dotnet/roslyn-analyzers to provide a public parameterless constructor if a SafeHandle provides a non-public constructor.

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Pull request merged**